### PR TITLE
Rewriting matmul op verifier to be compatible with torch's matmul

### LINKED ISSUE #717: TTIR Enforces that Matmul operands have same rank
As has been discussed, we should probably get rid of this and verify it in the TTNN dialect if at all. This is blocking resnet (tensorflow) ad the classification head has rank-2 weights and a rank-3 activation, which we should be able to perform anyway.

The following example should be legal: 
```
// RUN: ttmlir-opt --ttir-load-system-desc --ttir-implicit-device --ttir-layout --convert-ttir-to-ttnn %s | FileCheck %s
#any_device_tile = #tt.operand_constraint<dram|l1|tile|any_device_tile>
// CHECK: #[[TILED_LAYOUT:.*]] = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x4x!tt.tile<32x32, bf16>, #dram>, interleaved>
module attributes {} {
  func.func @forward(%arg0: tensor<1x64x128xbf16>, %arg1: tensor<128x96xbf16>) -> tensor<1x64x96xbf16> {
    %0 = tensor.empty() : tensor<1x64x96xbf16>
    // CHECK: %[[C:.*]] = "ttnn.matmul"[[C:.*]]
    %1 = "ttir.matmul"(%arg0, %arg1, %0) <{operand_constraints = [#any_device_tile, #any_device_tile, #any_device_tile]}> : (tensor<1x64x128xbf16>, tensor<128x96xbf16>, tensor<1x64x96xbf16>) -> tensor<1x64x96xbf16>
    return %1 : tensor<1x64x96xbf16>
  }
}
```

It will fail with:
```
test/ttmlir/Dialect/TTNN/simple_matmul.mlir:8:10: error: 'ttir.matmul' op Input A and B must have the same rank
    %1 = "ttir.matmul"(%arg0, %arg1, %0) <{operand_constraints = [#any_device_tile, #any_device_tile, #any_device_tile]}> : (tensor<1x64x128xbf16>, tensor<128x96xbf16>, tensor<1x64x96xbf16>) -> tensor<1x64x96xbf16>
```

This message comes from `lib/Dialect/TTIR/IR/TTIROps.cpp` line 422

### PR DESCRIPTION
**Pull request by sdjordjevicTT**
I am rewriting the Matmul verifier on both the TTIR and TTNN dialects per [Torch's](https://pytorch.org/docs/stable/generated/torch.matmul.html) and [NumPy](https://numpy.org/doc/stable/reference/generated/numpy.matmul.html) specifications.

I copy\pasted verifier from TTIR to TTNN as per TTNN public docs, it should support torch like Matmul:
https://docs.tenstorrent.com/ttnn/latest/ttnn/ttnn/matmul.html

I've added multiple tests to verify the TTIR and TTNN Matmul ops with the new verifier, covering both positive and negative cases. I didn't include any of the silicon tests because several examples are failing in TTNN. I will open bugs on the TTNN side to keep track of these failures.

Resolves #717 and #333.

### CODE REVIEW

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 1]
```diff
@@ -409,34 +414,149 @@ ::mlir::LogicalResult mlir::tt::ttir::MatmulOp::verify() {
...
+
+  if (inputBType.getRank() > 1) {
+    expectedOutputShape.push_back(inputBShape[inputBShape.size() - 1]);
   }
+
+  // Check the case of a vector-vector product. At this moment we don't support
+  // scalars in IR, hence check that the output is at least 1D tensor of size 1.
+  if (expectedOutputShape.size() == 0) {
+    if (outputType.getRank() < 1) {
+      return emitOpError("Scalar output is not supported, output must be at "
```
Comment (sdjordjevicTT) [inline lib/Dialect/TTIR/IR/TTIROps.cpp:529]: In the past I was opposed to having scalar as a part of IR(like we can pack scalars in one element array), maybe we should reconsider that now? :)

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 2]
```diff
@@ -409,34 +414,149 @@ ::mlir::LogicalResult mlir::tt::ttir::MatmulOp::verify() {
...
       inputBShape[inputBShape.size() - 2]) {
-    return emitOpError("Input A and B must have matching inner dimensions");
+    return emitOpError(
+        "Input A[-1](" + std::to_string(inputAShape[inputAShape.size() - 1]) +
+        ") and B[-2](" + std::to_string(inputBShape[inputBShape.size() - 2]) +
+        ") must have matching inner dimensions");
   }
-  if (outputShape[outputShape.size() - 2] !=
-      inputAShape[inputAShape.size() - 2]) {
-    return emitOpError("Output must have the same number of rows as input A");
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/IR/TTIROps.cpp:433]: Sometimes for these kinds of checks, it can be simpler to normalize the rank of both inputs first:
```c++
while (inputAShape.size() < inputBShape.size) inputAShape.insert(..., 1);
while (inputBShape.size() < inputAShape.size) inputBShape.insert(..., 1);
```

Now you can compare respective dims in lockstep with each other and assert for equality or if one or the other is `1` for legal bcastable dims.  It's then trivial to form an output shape.

File: lib/Dialect/TTIR/IR/TTIROps.cpp (hunk 1 above)
Comment (nsmithtt) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:529]: Yeah it's kind of a PITA, maybe we can relax and support 0 rank tensors, it's another special case we need to handle when constructing one and passing into the runtime, but maybe it's not a big deal.

File: lib/Dialect/TTIR/IR/TTIROps.cpp (hunk 2 above)
Comment (sdjordjevicTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:433]: At this particular place, I am checking the inner dimensions. After that, I created the output expectedOutputShape which later I checked against the provided output shape.

I tried to reuse the llvm's OpTrait::util::getBroadcastedShape to get the broadcast check and form the output expectedOutputShape. Maybe I can rewrite the code for checking the broadcast more simpler without it, but not sure.

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 3]
```diff
@@ -409,34 +414,149 @@ ::mlir::LogicalResult mlir::tt::ttir::MatmulOp::verify() {
...
-  if (inputAShape.size() < 2) {
-    return emitOpError("Input A must be at least a 2D tensor");
+
+  llvm::ArrayRef<int64_t> outputShape = outputType.getShape();
+  llvm::SmallVector<int64_t> inputAShape, inputBShape;
+
+  // Assemble the shape of the input tensor A
+  for (int i = 0; i < inputAType.getRank(); i++) {
+    inputAShape.push_back(inputAType.getShape()[i]);
   }
```
[the remark this answers is not in the document]
Comment (sdjordjevicTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:424]: Very nice! Will fix this. :)

File: lib/Dialect/TTIR/IR/TTIROps.cpp (hunk 1 above)
Comment (sdjordjevicTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:520]: Created the following issue to track this:
https://github.com/tenstorrent/tt-mlir/issues/767

**Review by nsmithtt — APPROVED** [review]

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 4]
```diff
@@ -409,34 +414,140 @@ ::mlir::LogicalResult mlir::tt::ttir::MatmulOp::verify() {
...
+  // Verify that the input A is at least 1D tensor
+  if (inputAType.getRank() < 1) {
+    return emitOpError("Input A must be at least a 1D tensor");
   }
-  if (inputAShape.size() != outputShape.size()) {
-    return emitOpError("Input A and B must have the same rank as the output");
+
+  // Verify that the input B is at least 1D tensor
+  if (inputBType.getRank() < 1) {
+    return emitOpError("Input B must be at least a 1D tensor");
```
Comment (rpavlovicTT) [inline lib/Dialect/TTIR/IR/TTIROps.cpp:442]: Maybe move these checks to be first in line.

**Review by rpavlovicTT — APPROVED** [review]

File: lib/Dialect/TTIR/IR/TTIROps.cpp (hunk 4 above)
Comment (sdjordjevicTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:442]: Agree, thanks for suggestion!

File: lib/Dialect/TTIR/IR/TTIROps.cpp (hunk 2 above)
Comment (sdjordjevicTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:433]: I created a join function in utils.h, it helps a lot to shorten the code and makes it much more readable now.
